File: server/internal/dto/response/checkout.go
package response

import (
	"debian-ecommerce/internal/data/entity"
	"time"
)

type PreviewCheckout struct {
	Items []PreviewItem `json:"selected_items"`
	Totals Totals `json:"totals"`
}

type PreviewItem struct {
	ID          uint                        `json:"id"`
	SKUID       uint                        `json:"sku_id"`
	SKUCode     string                      `json:"sku_code"`
	Product     ProductSnapshot             `json:"product"`
	Variants    []entity.VariantCombination `json:"variants"`
	Price       PriceSnapshot               `json:"price"`
	Quantity    int                         `json:"quantity"`
	Stock       int                         `json:"stock"`
	SubTotal    float64                     `json:"subtotal"`
}

type PromotionSummary struct {
	ID      uint      `json:"id"`
	Name    string    `json:"name"`
	EndDate time.Time `json:"end_date"`
	DiscountType      entity.DiscountType    `json:"discount_type"`
	DiscountValue     float64                `json:"discount_value"`
	MinimumOrderValue float64                `json:"minimum_order_value"`
	MaximumDiscount   float64                `json:"maximum_discount"`
	UsageLeft        int                    `json:"usage_left"`
	VoucherCode       *string                `json:"voucher_code"`
}

func ToPromoSummary(p entity.Promotion) PromotionSummary {
	return PromotionSummary{
		ID: p.ID,
		Name: p.Name,
		EndDate: p.EndDate,
		DiscountType: p.DiscountType,
		DiscountValue: p.DiscountValue,
		MinimumOrderValue: p.MinimumOrderValue,
		MaximumDiscount: p.MaximumDiscount,
		UsageLeft: p.UsageLimit - p.UsedCount,
		VoucherCode: p.VoucherCode,
	}
}

File: client/src/types/checkout.ts
import { ItemPrice, ItemVariant, ProductPreview } from "./cart";

export interface CheckoutItem {
  id: number;
  sku_id: number;
  sku_code: string;
  product: ProductPreview;
  variants: ItemVariant[];
  price: ItemPrice;
  quantity: number;
  stock: number;
  subtotal: number;
}

export interface ShippingOption {
  name: string;
  code: string;
  service: string;
  description: string;
  cost: number;
  etd: string;
}

export interface ShippingOptionTemp {
  name: string;
  code: string;
  service: string;
  description: string;
  cost: number;
  etd: string;
  id: string;
}

export interface Promotion {
  id: number;
  name: string;
  end_date: string;
  discount_type: "percentage" | "amount";
  discount_value: number;
  minimum_order_value: number;
  maximum_discount: number;
  usage_left: number;
  voucher_code: string | null;
}

export interface Totals {
  payment_method?: string;
  subtotal: number;
  discount_amount: number;
  shipping_cost: number;
  grand_total: number;
}

export interface CheckoutPreviewResponse {
  selected_items: CheckoutItem[];
  totals: Totals;
}